_posts/2019-08-08-hacking-for-cruise-control.md

```diff
@@ -0,0 +1,227 @@
+```yaml
+apiVersion: kafka.strimzi.io/v0alpha1
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    image: <kafka-image-name>
+    version: 2.0.1
+    replicas: 3
+    listeners:
+      plain: {}
+      tls: {}
+    config: 
+      metric.reporters: "com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter"
+      offsets.topic.replication.factor: 1
+      transaction.state.log.replication.factor: 1
+      transaction.state.log.min.isr: 1
+      log.message.format.version: "2.0.1"
+    storage:
+      type: ephemeral
+  zookeeper:
+    replicas: 3
+    storage:
+      type: ephemeral
+  entityOperator:
+    userOperator: {}
+    topicOperator: {}
+```
+Apply to cluster:
+
+```bash
+oc apply -f kafka.yaml
+```
+
+After recieving metric objects directly from Kafka, the cruise control metric reporter will filter out undesired metric objects, and store the remaining in a concurrent hash map. Then at periodic intervals designated by the user, the reporter will loop through the concurrent hash map and wrap each metric object's value into a Cruise Control Metric object. The reporter will add information like the metric type, time of when the metric value was recorded, and the broker id of where the metric was taken from. Finally, then Cruise Control metric object is formatted into a byte array and stored back into Kafka within a topic designated for cruise control metics.
+
+# Accessing Zookeeper
+
+Be default, communication between Zookeeper and all other Strimzi components must be encrytped. Using a hack by Jakub Scholz, one can create a Kubernetes service which reuses the Strimzi certs to pipe unecrypted Cruise Control traffic through to Zookeeper. This removes the need to include traffic encryting sidecars in the Cruise Control deployment.
+
+```
+kubectl apply -f https://gist.githubusercontent.com/scholzj/6cfcf9f63f73b54eaebf60738cfdbfae/raw/068d55ac65e27779f3a5279db96bae03cea70acb/zoo-entrance.yaml
+```
+
+# Containerizing and deploying Cruise Control
+
+Create a Cruise Control image
+
+```Dockerfile
+FROM centos:7
+
+ENV CC_VERSION=2.0.57
+ENV JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk
+
+RUN yum -y install git java-1.8.0-openjdk-devel && \
+    yum clean all -y
+
+RUN git clone --branch ${CC_VERSION} https://github.com/linkedin/cruise-control.git
+
+WORKDIR cruise-control
+
+RUN ./gradlew jar copyDependantLibs
+
+RUN sed -i 's/bootstrap.servers=.*/bootstrap.servers=my-cluster-kafka-bootstrap:9092/g' config/cruisecontrol.properties \
+    && sed -i 's/zookeeper.connect=.*/zookeeper.connect=zoo-entrance:2181/g' config/cruisecontrol.properties \
+    && sed -i 's/capacity.config.file=.*/capacity.config.file=config\/capacityJBOD.json/g' config/cruisecontrol.properties \
+    && sed -i 's/sample.store.topic.replication.factor=.*/sample.store.topic.replication.factor=1/g' config/cruisecontrol.properties 
```

Comment:
Instead of "building" the configuration properties file here during the image build, is it possible to use a ConfigMap where to put the configuration and mounting that one as a volume in the CC deployment?

Reply:
Definitely possible, in fact, it would be the proper way to do this. For the blog post, I was hoping to keep the instructions down to a minimum, but I do see the utility in doing this (quick prototyping, testing, etc). You think its worth making this change?